File: leetcode2/CreateSortedArraythroughInstructions/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2020 by iamslash 

from typing import List

# 5252ms 80.50% 28MB 5.03%
# fenwick tree
# O(NlgN) O(N)
class Solution:
    def createSortedArray(self, I: List[int]) -> int:
        max_val, mod = max(I), 1000000007
        data = [0] * (max_val + 1)
        def lsb(x):
            return x & -x
        def add(x):
            while x <= max_val:
                data[x] += 1
                x += lsb(x)
        def sum(x):
            ans = 0
            while x > 0:
                ans += data[x]
                x -= lsb(x)
            return ans
        ans = 0
        for i, a in enumerate(I):
            ans = (ans + min(sum(a-1), i - sum(a))) % mod
            add(a)
        return ans
    
if __name__ == "__main__":
    sln = Solution()
    print(sln.createSortedArray([1,5,6,2])) # 1
    print(sln.createSortedArray([1,2,3,6,5,4])) # 3
    print(sln.createSortedArray([1,3,3,3,2,4,2,1,2])) # 4

File: leetcode2/CreateSortedArraythroughInstructions/a.cpp
/* Copyright (C) 2020 by iamslash */

#include <cstdio>
#include <vector>

using namespace std;

// 488ms 79.51% 143.6MB 5.43%
// binary indexed tree
// O(NlgN) O(N)
#define MOD 1000000007
class Solution {
 public:
  vector<int> data = vector<int>(100001);
  int createSortedArray(vector<int>& A) {
    int n = A.size(), ans = 0;
    for (int i = 0; i < n; ++i) {
      ans = (ans + min(sum(A[i]-1), i - sum(A[i]))) % MOD;
      add(A[i]);
    }
    return ans;
  }
  int lsb(int x) {
    return x & -x;
  }
  void add(int x) {
    while (x < 100001) {
      data[x]++;
      x += lsb(x);
    }
  }
  int sum(int x) {
    int ans = 0;
    while (x > 0) {
      ans += data[x];
      x -= lsb(x);
    }
    return ans;
  }
};
